client/src/lib.rs

```diff
@@ -87,20 +112,40 @@ pub fn read_response_header(buffer: &[u8]) -> TssResult<(RespHeader, usize)> {
 }
 
 /// Unmarshals any response sessions.
-pub fn read_response_sessions(sessions: &CmdSessions, buffer: &mut UnmarshalBuf) -> TssResult<()> {
-    for session in sessions {
-        // TODO: Support parameter decryption.
-        let auth = TpmsAuthResponse::try_unmarshal(buffer)?;
-        session.validate_auth_response(&auth)?;
-    }
+pub fn read_response_sessions<
+    X: Session,
+    Y: Session,
+    Z: Session,
+    AA: AuthorizationArea<X, Y, Z>,
+>(
+    sessions: &AA,
+    buffer: &mut UnmarshalBuf,
+) -> TssResult<()> {
+    let (s1, s2, s3) = sessions.decompose_ref();
+    let Some(s1) = s1 else { return Ok(()) };
+    let auth = TpmsAuthResponse::try_unmarshal(buffer)?;
+    s1.validate_auth_response(&auth)?;
+    let Some(s2) = s2 else { return Ok(()) };
+    let auth = TpmsAuthResponse::try_unmarshal(buffer)?;
+    s2.validate_auth_response(&auth)?;
+    let Some(s3) = s3 else { return Ok(()) };
+    let auth = TpmsAuthResponse::try_unmarshal(buffer)?;
+    s3.validate_auth_response(&auth)?;
     Ok(())
 }
 
 /// Runs a command with provided handles and sessions.
-pub fn run_command_with_handles<CmdT, T>(
+pub fn run_command_with_handles<
+    CmdT,
+    T,
+    X: Session,
+    Y: Session,
+    Z: Session,
+    AA: AuthorizationArea<X, Y, Z>,
+>(
     cmd: &CmdT,
     cmd_handles: CmdT::Handles,
-    cmd_sessions: CmdSessions,
+    cmd_sessions: AA,
```

Comment:
So, how this reads to me provides no demand on the caller to provide a specific number of authorization sessions based on the command. I don't see how this forces us to provide the right number of sessions, because everything just devolves to an `AuthorizationArea<X, Y, Z>` (which can be 0-3 sessions in all cases).

Reply:
If you want to get some protection provided by the compiler, you'll have to mark up the commands themselves.

Reply:
That is true, the moment we have commands, they must be annotated with which  traits they support, but we only need to pick one. After all, the information must come from somewhere. The key thing here is from user perspective, they just pass tuple of sessions, where all types are hidden, business logic implementer perspective it is all a `From`/`Into` cast (and `TryFrom`/`TryInto` when needed to to the other way around) But at the end of day, at the protocol level, this information will be thrown away, because TPM specifications does not preserve this information in the protocol itself.

Reply:
I'm not sure I follow your plan here, but we can keep heading down this path to see where it takes us.

---

We do have [some commands](https://github.com/tpm-rs/tpm-rs/blob/main/base/src/commands.rs), and at least one of those commands say "I don't accept *any* sessions", so I think it would be good to practice on that. I'm fine with it being in a separate PR to keep things small though.

From my glance at the TPM2 spec, I see the following of our commands w.r.t. what they say about sessions:

*   `GetCapability` -> 0-3 sessions (allows optional sessions, requires none)
*   `Startup` -> 0 sessions (disallows optional sessions, requires none)
*   `Shutdown` -> 0-3 sessions (allows optional sessions, requires none)
*   `PcrRead` -> 0-3 sessions (allows optional sessions, requires none)

---

I'll say again, I'm super worried that this is going to turn into template madness. I'm still feeling a slight opposition to all this static dispatching... I worry it's going to build a confusing API and lead to code bloat, I'm not sure that's a trade I'm happy about.

Anyways, I'm fine continuing down this path for the time being as an experiment though. I just would feel wrong not saying it somewhere in the PR review for tracking purposes.